server/src/main/java/org/elasticsearch/action/bulk/IncrementalBulkService.java

```diff
@@ -0,0 +1,194 @@
+        private BulkRequest bulkRequest = null;
+
+        private Handler(Client client, @Nullable String waitForActiveShards, @Nullable TimeValue timeout, @Nullable String refresh) {
+            this.client = client;
+            this.waitForActiveShards = waitForActiveShards != null ? ActiveShardCount.parseString(waitForActiveShards) : null;
+            this.timeout = timeout;
+            this.refresh = refresh;
+            createNewBulkRequest(BulkRequest.IncrementalState.EMPTY);
+        }
+
+        public void addItems(List<DocWriteRequest<?>> items, Releasable releasable, Runnable nextItems) {
+            if (topLevelFailure != null) {
+                assert releasables.isEmpty();
+                assert bulkRequest == null;
+                if (globalFailure == false) {
+                    addItemLevelFailures(items);
+                }
+                Releasables.close(releasable);
+                nextItems.run();
+            } else {
+                internalAddItems(items, releasable);
+
+                if (shouldBackOff()) {
+                    client.bulk(bulkRequest, ActionListener.runAfter(new ActionListener<>() {
+
+                        private final boolean isFirstRequest = incrementalRequestSubmitted == false;
+
+                        @Override
+                        public void onResponse(BulkResponse bulkResponse) {
+                            responses.add(bulkResponse);
+                            releaseCurrentReferences();
+                            createNewBulkRequest(bulkResponse.getIncrementalState());
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            handleBulkFailure(isFirstRequest, e);
+                        }
+                    }, nextItems));
+                    incrementalRequestSubmitted = true;
+                } else {
+                    nextItems.run();
+                }
+            }
+        }
+
+        private boolean shouldBackOff() {
+            // TODO: Implement Real Memory Logic
+            return bulkRequest.requests().size() > 16;
+        }
+
+        public void lastItems(List<DocWriteRequest<?>> items, Releasable releasable, ActionListener<BulkResponse> listener) {
+            if (globalFailure) {
+                assert releasables.isEmpty();
+                assert bulkRequest == null;
+                releasable.close();
+                listener.onFailure(topLevelFailure);
+            } else {
+                internalAddItems(items, releasable);
+
+                client.bulk(bulkRequest, ActionListener.runAfter(new ActionListener<>() {
+
+                    private final boolean isFirstRequest = incrementalRequestSubmitted == false;
+
+                    @Override
+                    public void onResponse(BulkResponse bulkResponse) {
+                        responses.add(bulkResponse);
+                        releaseCurrentReferences();
```

Comment:
It would be nice to null out `bulkRequest` here. That would also allow us to assert that `bulkRequest != null` in add/lastItems to verify the interaction. I think doing it in `releaseCurrentReferences` could be beneficial.